features/device_key/source/DeviceKey.cpp

```diff
@@ -0,0 +1,292 @@
+{
+    //KDF in counter mode implementation as described in Section 5.1
+    //of NIST SP 800-108, Recommendation for Key Derivation Using Pseudorandom Functions
+    int ret;
+    size_t counter = 0;
+    char separator = 0x00;
+    mbedtls_cipher_context_t ctx;
+    unsigned char output_len_enc[ 4 ] = {0};
+    unsigned char counter_enc[ 1 ] = {0};
+
+    DEVKEY_WRITE_UINT32_LE(output_len_enc, ikey_type);
+
+    mbedtls_cipher_type_t mbedtls_cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
+    if (DEVICE_KEY_32BYTE == ikey_size) {
+        mbedtls_cipher_type = MBEDTLS_CIPHER_AES_256_ECB;
+    }
+
+    const mbedtls_cipher_info_t *cipher_info = mbedtls_cipher_info_from_type(mbedtls_cipher_type);
+
+    mbedtls_cipher_init(&ctx);
+    ret = mbedtls_cipher_setup(&ctx, cipher_info);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    do {
+
+        ret = mbedtls_cipher_cmac_starts(&ctx, (unsigned char *)ikey_buff, ikey_size * 8);
+        if (ret != 0) {
+            goto finish;
+        }
+
+        DEVKEY_WRITE_UINT8_LE(counter_enc, (counter+1));
+
+        ret = mbedtls_cipher_cmac_update(&ctx, (unsigned char *)counter_enc, sizeof(counter_enc));
+        if (ret != 0) {
+            goto finish;
+        }
+
+        ret = mbedtls_cipher_cmac_update(&ctx, isalt, isalt_size);
+        if (ret != 0) {
+            goto finish;
+        }
+
+        ret = mbedtls_cipher_cmac_update(&ctx, (unsigned char *)&separator, sizeof(char));
+        if (ret != 0) {
+            goto finish;
+        }
+
+        ret = mbedtls_cipher_cmac_update(&ctx, (unsigned char *)&output_len_enc, sizeof(output_len_enc));
+        if (ret != 0) {
+            goto finish;
+        }
+
+        ret = mbedtls_cipher_cmac_finish(&ctx, output + (DEVICE_KEY_16BYTE * (counter)));
+        if (ret != 0) {
+            goto finish;
+        }
+
+        counter++;
+
+    } while (DEVICE_KEY_16BYTE * counter < ikey_type);
+
+finish:
+    mbedtls_cipher_free( &ctx );
+
+    if (DEVICEKEY_SUCCESS != ret) {
+        return DEVICEKEY_ERR_CMAC_GENERIC_FAILURE;
+    }
+
+    return DEVICEKEY_SUCCESS;
+}
+
+int DeviceKey::generate_key_by_trng(uint32_t *output, size_t& size)
+{
+#if defined(DEVICE_TRNG)
+    size_t in_size;
+    size_t ongoing_size;
+    size_t final_size;
+    trng_t trng_obj;
+    int ret = DEVICEKEY_SUCCESS;
+    unsigned char *pBuffer = (unsigned char *)output;
+
+    memset(output, 0, size);
+
+    if (DEVICE_KEY_16BYTE > size) {
+        return DEVICEKEY_BUFFER_TO_SMALL;
+    } else if (DEVICE_KEY_16BYTE <= size && DEVICE_KEY_32BYTE > size) {
+        size = DEVICE_KEY_16BYTE;
+    } else {
+        size = DEVICE_KEY_32BYTE;
+    }
+
+    trng_init(&trng_obj);
+
+    final_size = 0;
+    in_size = size;
+    while (true) {
+
+        ongoing_size = 0;
+        ret = trng_get_bytes(&trng_obj, (unsigned char *)pBuffer, in_size, &ongoing_size);
+        final_size += ongoing_size;
+        if (DEVICEKEY_SUCCESS != ret) {
```

Comment:
The TRNG API uses `-1` as the return code in case of a generic failure. Please change.

Reply:
@hanno-arm 
Sure, but i prefer to stay with checking the not success if it is fine by you.

Reply:
@yossi2le While the usage of `0` as indicating success is quite standard, it is error-prone to mix error code constants from different modules. What if suddently `DEVICEKEY_SUCCESS` or the semantics of `trng_get_bytes` change?

Reply:
I think I had a look at an old version here -- the present one is fine by me.